1-Draft/RFC-Support for .NET extension methods.md

```diff
@@ -0,0 +1,82 @@
+
+This RFC proposes changes to language syntax and runtime binding to enable the calling of generic .NET methods, and finding appropriate extension methods based on the type of the extension methods first parameter.
+
+## Motivation
+
+Extension methods enable you to "add" methods to existing types without creating a new derived type, recompiling, or otherwise modifying the original type. Extension methods are a special kind of static method, but they are called as if they were instance methods on the extended type. For client code written in PowerShell, there would be no apparent difference between calling an extension method and the methods that are actually defined in a type.
+
+There are lots of useful extension methods in .NET already written, many of which are in the area of processing sets of objects and where the expressiveness/performance ratio, from a PowerShell perspective, is quite favorable.
+
+### Goals
+
+Simplicity of use
+-	Should be as easy as possible to call from PowerShell.
+    -	Type parameters will be deduced to the most specific type possible
+```powershell
+    # System.Linq.Enumerable.Select[TInput, TResult] takes two generic type parameters
+    using namespace System.Linq    
+    $a.Select({param([int] $i) $i.ToString()}) # [int,object] will be deduced.
+
+    [func[int,string]] $f = {param([int]$i) ($i * 2).ToString()}
+    $a.Select($f) # [int,string] will be deduced.
+    
+    $a.Select(param($i) $i * 2) #  [object,object] will be deduced.
+```
+
+Simplication of 'Magic' methods
+The `Foreach` and `Where` methods that are added to collections are a hack which could be solved more elegantly by making extension methods of the EnumerableOps methods.
+By doing this, they wouldn't need to be special cased. Normal method resolution would make it work, and would also make `List[T].Foreach` not hide the extension methods.
+
+```csharp
+public static class EnumerableOps {
+    static IEnumerable<T> Foreach<T>(this IEnumerable<T> c, ScriptBlock sb){}
+    static IEnumerable<T> Foreach<T>(this IEnumerable<T> c, ScriptBlock sb, WhereOperatorSelectionMode mode){}
+    static IEnumerable<T> Foreach<T>(this IEnumerable<T> c, ScriptBlock sb, WhereOperatorSelectionMode mode, int count){}
+    static IEnumerable<T> Foreach<T>(this IEnumerable<T> c, object expression, object[] args){}
+
+    static IEnumerable<T> Where<T>(this IEnumerable<T> c, ScriptBlock sb){}
+    static IEnumerable<T> Where<T>(this IEnumerable<T> c, ScriptBlock sb, WhereOperatorSelectionMode mode){}
+    static IEnumerable<T> Where<T>(this IEnumerable<T> c, ScriptBlock sb, WhereOperatorSelectionMode mode, int count){}
+}
+
+```
+
+
+## Specification
+
+Extension methods are choosen from all methods on all loaded types where the methods has the ExtensionAttribute.
+From these, the resulting set is limited to the types whose namespace are in use in the context where the call is made.
+From these, a best fit is selected based on parameters.
```

Comment:
Reducing the search space is desirable for two Speed and Accuracy.

I'm not sure how slow this is going to be, searching all loaded types for extension methods is something that all other languages do at _compile_ time, once.  Would PowerShell resolve extension methods at parse time, or at run time?

I think that extension methods in compiled languages are _safe_ **because** we _must_ explicitly link libraries and add using statements before the extension methods show up, and even then we must sometimes resolve collisions definitively at compile-time.

I would almost rather this only work with the `using assembly ___` and/or `using namespace ___` syntax...